Do not publish form/case changes on delete domain

millerdev opened the pull request:
Domain deletions have been causing high load, resulting in pillow backlogs and general HQ performance degradation.

This PR moves Elasticsearch deletions to a synchronous function that deletes all documents for the domain in `form`, `case`, and `case_search` indexes instead of propagating them through the change feed. Form and case deletion events are not sent via repeaters, so that is not a concern.

One advantage of this new way of deleting documents in ES is that all documents for the domain in the form and case indexes are deleted each time `delete_domain` is run. So documents in ES will not need to be handled specially (with `delete_es_docs_for_domain`) if a command needs to be killed or dies for some other reason while deleting forms or cases. Simply run `delete_domain` again until it succeeds.

## Safety Assurance

### Safety story

Only affects domain deletions.

### Automated test coverage

Yes.

### QA Plan

No.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

millerdev committed:
Do not publish form/case changes on delete domain

dannyroberts commented:
My mental model so far has been that the point of publishing the deletes is that it causes those deletes to get propagated to Elasticsearch. We certainly do want that data to get deleted from elasticsearch as part of this processes. I don't know if it gets deleted by some other means, but my assumption was that it doesn't

millerdev committed:
Bulk delete cases and forms in Elasticsearch

dannyroberts commented:
Some thoughts on things that might have to be addressed (may or may not already be):

- There are multiple case indices in ES—do they all get deleted?
- As you noted, does this change anything on the repeater side? Even if we're okay with that, we should know whether it does or not

These might also already be caught by tests. This isn't necessarily an exhaustive list, but it's what came to mind.

AmitPhulera commented on corehq/apps/domain/deletion.py:
```diff
@@ -188,15 +190,17 @@ def delete_all_cases(domain_name):
     for chunk in chunked(case_ids, 1000, list):
-        CommCareCase.objects.hard_delete_cases(domain_name, chunk)
+        CommCareCase.objects.hard_delete_cases(domain_name, chunk, publish_changes=False)
+        case_adapter.bulk_delete(chunk)
```
This should also delete from `case_search` index if case_search is enabled on domain.

AmitPhulera commented:
The approach looks better than what we were doing before. I have the same suggestion to what @dannyroberts made about deleting from case-search index as well.

> Do we guarantee that form and case deletion events will be sent via repeaters when a domain is being deleted?

When we are deleting a domain does it not mean that the client does not need data anymore, I don't think it would make sense to create repeat records for delete request? From my understanding the repeat records are linked to the django signal and all the deletes will create repeat records which in turn will clog the the repeat record queue. Not sure what the SLA for this would be but it feels okay to me not to send out the deletes via repeaters if domain is to be deleted.

millerdev commented:
@AmitPhulera, yes I agree with all of your points about repeaters. The question was posed to prompt confirmation and potentially highlight similar side effects that might be affected by this change.

millerdev committed:
Guard global patches against patch.stopall()

millerdev committed:
Replace patch.stopall() with explicit patch stops

Not strictly necessary, but seems safer. Also, it's difficult to track down if patch.stopall() causes some other test to fail.

AmitPhulera commented on corehq/apps/domain/deletion.py:
```diff
@@ -219,6 +223,20 @@ def iter_ids(model_class, field, domain, chunk_size=1000):
+            stream=stream,
+        )
+        for chunk in chunked(doc_ids, 1000, list):
+            query.adapter.bulk_delete(chunk)
```
If deletes fail for some reason, the expectation here is that we will run the deletion script again, right?

millerdev replied:
Correct.
